modules/libsbc/sbc.c

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <zephyr/libsbc/sbc.h>
+
+#if defined(CONFIG_LIBSBC_ENCODER)
+
+int sbc_setup_encoder(struct sbc_encoder *encoder, struct sbc_encoder_init_param *param)
+{
+	SBC_ENC_PARAMS *encoder_params;
+
+	if (encoder == NULL) {
+		return -1;
+	}
+
+	memset(encoder, 0, sizeof(struct sbc_encoder));
+
+	encoder_params = &encoder->sbc_encoder_params;
+
+	encoder_params->s16ChannelMode = (int16_t)param->ch_mode;
+	encoder_params->s16NumOfSubBands = (int16_t)param->subband;
+	if (!encoder_params->s16NumOfSubBands) {
+		return -EINVAL;
+	}
+	encoder_params->s16NumOfBlocks = (int16_t)param->blk_len;
+	if (!encoder_params->s16NumOfBlocks) {
+		return -EINVAL;
+	}
+	encoder_params->s16AllocationMethod = (int16_t)param->alloc_mthd;
+	encoder_params->s16NumOfChannels = param->ch_num;
+	if (!encoder_params->s16NumOfChannels) {
+		return -EINVAL;
+	}
+
+	switch (param->samp_freq) {
+	case 16000u:
+		encoder_params->s16SamplingFreq = 0;
+		break;
+	case 32000u:
+		encoder_params->s16SamplingFreq = 1;
+		break;
+	case 44100u:
+		encoder_params->s16SamplingFreq = 2;
+		break;
+	case 48000u:
+		encoder_params->s16SamplingFreq = 3;
+		break;
+	default:
+		return -1;
+	}
+
+	encoder_params->u16BitRate = param->bit_rate;
+
+	SBC_Encoder_Init(encoder_params);
+
+	if (encoder_params->s16BitPool < param->min_bitpool) {
+		/* need to increase the `param->bit_rate` */
+		return -EINVAL;
+	} else if (encoder_params->s16BitPool > param->max_bitpool) {
+		/* need to decrease the `param->bit_rate` */
+		return -EOVERFLOW;
+	}
+
+	return 0;
+}
+
+/**
+ * Encode a SBC frame
+ */
+uint32_t sbc_encode(struct sbc_encoder *encoder, const void *in_data, void *out_data)
+{
+	uint32_t ret;
+
+	if ((encoder == NULL) || (in_data == NULL) || (out_data == NULL)) {
+		return -1;
```

Comment:
This function returns `uint32_t` so it doesn't make sense to try to return a negative value.